File: src/main/java/co/gatedaccess/web/SecureInterceptor.java
package co.gatedaccess.web;

import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier;
import com.google.api.core.ApiFuture;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseToken;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;

import javax.annotation.Nonnull;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

@Component
public class SecureInterceptor implements HandlerInterceptor {
    private static final Logger log = Logger.getLogger(SecureInterceptor.class.getSimpleName());


    @Override
    public boolean preHandle(HttpServletRequest request, @Nonnull HttpServletResponse response, @Nonnull Object handler) throws Exception {
        String idToken = request.getHeader("Authorization");

        if (idToken != null) {
            try {
                idToken = idToken.split(" ")[1]; //Remove Bearer prefix

                // Running on dev, skip token decode
                if (request.getRequestURL().toString().startsWith("http://localhost:8080")) {
                    System.out.println("Bearer is: " + idToken);
                    request.setAttribute("user", idToken);
                    return true;
                }

                ApiFuture<FirebaseToken> tokenAsync = FirebaseAuth.getInstance().verifyIdTokenAsync(idToken, true);
                FirebaseToken firebaseToken = tokenAsync.get(5, TimeUnit.SECONDS);
                request.setAttribute("user", firebaseToken.getUid());
                return true;
            } catch (Exception e) {
                log.info(e.getLocalizedMessage());
            }
        }

        response.setContentType("application/json");
        response.sendError(401, "Unauthorized user");
        return false;
    }
}

File: src/main/kotlin/co/gatedaccess/web/http/interceptor/SecureInterceptor.kt
package co.gatedaccess.web.http.interceptor

import co.gatedaccess.web.data.model.UserType
import co.gatedaccess.web.data.repo.MemberRepo
import co.gatedaccess.web.data.repo.SecurityGuardRepo
import com.google.api.core.ApiFuture
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseToken
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.core.env.StandardEnvironment
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatusCode
import org.springframework.stereotype.Component
import org.springframework.web.context.WebApplicationContext
import org.springframework.web.servlet.HandlerInterceptor
import java.util.concurrent.TimeUnit
import javax.annotation.Nonnull

@Component
class SecureInterceptor(val context: WebApplicationContext) : HandlerInterceptor {

    override fun preHandle(
        request: HttpServletRequest,
        @Nonnull response: HttpServletResponse,
        @Nonnull handler: Any
    ): Boolean {

        var idToken = request.getHeader("Authorization")

        if (idToken != null) {
            try {
                idToken =
                    idToken.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1] //Remove Bearer prefix


                val userId : String
                val userType : UserType
                // Running on dev, skip token decode
                if (StandardEnvironment().activeProfiles[0] == "dev") {
                    userType = UserType.Admin // Manually update this to any type when running on dev
                    userId = idToken
                    println("Bearer is: $idToken")
                } else {
                    val tokenAsync : ApiFuture<FirebaseToken> = FirebaseAuth.getInstance()
                        .verifyIdTokenAsync(idToken, true)
                    val firebaseToken = tokenAsync[10, TimeUnit.SECONDS]
                    userType = UserType.valueOf(firebaseToken.claims["type"] as String)
                    userId = firebaseToken.uid
                }

                /** Get User model from DB and attach to request**/
                if(userType == UserType.Guard){
                    val guardRepo = context.getBean(SecurityGuardRepo::class.java)
                    val guard = guardRepo.findById(userId).orElseThrow()!!
                    request.setAttribute("user", guard)
                }else{
                    val memberRepo = context.getBean(MemberRepo::class.java)
                    val member = memberRepo.findById(userId).orElseThrow()!!
                    request.setAttribute("user", member)
                }
                return true
            } catch (e: Exception) {
                LoggerFactory.getLogger(this::class.java.simpleName).info(e.localizedMessage)
            }
        }
        response.sendError(HttpStatus.UNAUTHORIZED.value(), "User is not logged in")
        return false
    }

}
